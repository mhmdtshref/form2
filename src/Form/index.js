import React, { Component } from 'react'

class Form extends Component {
  constructor(props) {
    super(props)
    this.state = {
      values: {}
    }
  }

  setStateValue = (name, value) => {
    this.setState((prevState) => ({...prevState.values, [name]: value}), () => {
      console.log("STATE CHANGED:::", this.state);
    })
  };

  setEventToState = ({ target }) => {
    this.setStateValue([target.name], target.value)
  };

  renderTextField = (field, index) => {
    field.key = field.key ? field.key : index;
    if (field.onChange) {
      return <input
        type='text'
        name={field.name}
        onChange={(event) => { field.onChange(event); this.setEventToState(event) }}
        {...field}
      />
    } else {
      return <input
        type='text'
        name={field.name}
        onChange={this.setEventToState}
        {...field}
      />
    }
  };

  renderFields = (fields) => {
    return fields.map((field, index) => {
      switch (field.type) {
        case 'text':
          return this.renderTextField(field, index)
      }
    })
  };


  renderChildren = () => {
    const { setStateValue } = this;
    return React.Children.toArray(this.props.children).map((child, key) => React.cloneElement(child, { formstate: setStateValue, key }));
  };

  render() {
    return <form>
      { this.renderChildren() }
    </form>
  }
}

export default Form
