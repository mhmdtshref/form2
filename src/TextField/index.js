import React, { Component } from 'react'

class TextField extends Component {
  constructor(props) {
    super(props);
    this.state = {
      value: this.props.value || '',
    };
  }

  onChange = ({ target }) => {
    this.setState({ value: target.value }, () => {
      this.props.formstate(this.props.name, this.state.value);
      if(this.props.onChange) this.props.onChange();
    });
  };

  generateProps = (props) => {
    let newProps = {...props};
    delete newProps.onChange;
    delete newProps.formstate;
    return newProps;
  };

  componentDidMount() {
      this.props.formstate(this.props.name, this.state.value);
  }

    render() {
    return <input value={this.state.value} onChange={this.onChange} {...this.generateProps(this.props)} />;
  }
}

export default TextField;
