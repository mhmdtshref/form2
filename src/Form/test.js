import React from 'react';
import renderer from 'react-test-renderer';
import Form from './index';
import TextField from "../TextField";
import PasswordField from "../PasswordField";
import DatePicker from '../DatePicker';

describe('Form Test', () => {

  it('is truthy', () => {
    expect(Form).toBeTruthy()
  });

  it('Form with one field in it', () => {
    const form = renderer.create(<Form>
      <TextField name="username" />
    </Form>).toJSON();
    expect(form).toMatchSnapshot();
  });

  it('Form with multiple fields in it', () => {
    const form = renderer.create(<Form>
      <TextField name="username" placeholder="Username"/>
      <PasswordField name="password" placeholder="email" />
      <DatePicker name="birthday" />
    </Form>).toJSON();
    expect(form).toMatchSnapshot();
  });
});
