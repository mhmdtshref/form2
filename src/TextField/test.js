import React from 'react';
import renderer from 'react-test-renderer';
import TextField from './index';

describe('Password Field Test', () => {

  it('is truthy', () => {
    expect(TextField).toBeTruthy()
  });

  it('Test natural text field', () => {
    const textField = renderer.create(<TextField name="username" />).toJSON();
    expect(textField).toMatchSnapshot();
  });

  it('Test text field with placeholder property', () => {
    const textField = renderer.create(<TextField name="username" placeholder="Type your name..." />).toJSON();
    expect(textField).toMatchSnapshot();
  });

  it('Test text field with onChange function property', () => {
    const onChangeHandler = () => {
      console.log("Test console log!");
    };
    const textField = renderer.create(<TextField name="username" onChange={onChangeHandler} />).toJSON();
    expect(textField).toMatchSnapshot();
  });
});
