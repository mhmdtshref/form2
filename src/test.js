import React from 'react';
import renderer from 'react-test-renderer';
import Form from './Form';
import TextField from "./TextField";

describe('Check form text field', () => {

  it('is truthy', () => {
    expect(Form).toBeTruthy()
  });

  it('Field of type text', () => {
    const form = renderer.create(<Form>
      <TextField name="username" />
      <TextField name="password" />
    </Form>).toJSON();
    expect(form).toMatchSnapshot();
  });

  it('Field of type text with placeholder in props', () => {
    const form = renderer.create(<Form>
      <TextField name="username" placeholder="Username"/>
      <TextField name="email" placeholder="email" />
    </Form>).toJSON();
    expect(form).toMatchSnapshot();
  });

  it('Field of type text with onChange function in the props ', () => {
    const form = renderer.create(<Form>
      <TextField name="username" onChange={() => { console.log("Hello1"); }} />
      <TextField name="password" onChange={() => { console.log("Hello2"); }} />
    </Form>).toJSON();
    expect(form).toMatchSnapshot();
  });

});
