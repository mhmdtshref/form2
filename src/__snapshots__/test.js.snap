// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check form text field Field of type text 1`] = `
<form>
  <input
    name="username"
    onChange={[Function]}
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    value=""
  />
</form>
`;

exports[`Check form text field Field of type text with onChange function in the props  1`] = `
<form>
  <input
    name="username"
    onChange={[Function]}
    value=""
  />
  <input
    name="password"
    onChange={[Function]}
    value=""
  />
</form>
`;

exports[`Check form text field Field of type text with placeholder in props 1`] = `
<form>
  <input
    name="username"
    onChange={[Function]}
    placeholder="Username"
    value=""
  />
  <input
    name="email"
    onChange={[Function]}
    placeholder="email"
    value=""
  />
</form>
`;
